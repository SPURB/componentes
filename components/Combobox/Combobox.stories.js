import { storiesOf } from '@storybook/vue'
import Combobox from './Combobox'

storiesOf('Combobox', module)
  .add('normal', () => ({
    components: { Combobox },
    template:
    `
      <Combobox
        :options="options"
        :type="true"
      />
    `,
    data: () => ({
      options: [
        {
          title: 'Selecione um projeto',
          value: 0
        },
        {
          title: 'PIU Bairros do Tamandatue√≠',
          value: 1
        },
        {
          title: 'PIU Jurubatuba',
          value: 2
        },
        {
          title: 'PIU Setor Central',
          value: 3
        }
      ]
    })
  }))