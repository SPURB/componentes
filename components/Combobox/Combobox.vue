<template>
  <div class="select-options">
    <button
      class="select-options__toggler"
      :class="[typeBackground, typeDisabled]"
      @click.prevent="open = !open"
      data-cy="btn__options"
      :disabled="disabled"
    >
      <div>{{ selected }}</div>
      <i class="icon icon-expandir" :class="{ open }"></i>
    </button>
    <transition name="fade">
      <ul class="select-options__select" v-show="open">
        <li
          class="select-options__options"
          v-for="(option, index) in options"
          :value="option.value"
          :key="index"
        >
          <button
            @click.prevent="setSelectedValue(option)"
            data-cy="select__option"
          >
            {{ option.title }}
          </button>
        </li>
      </ul>
    </transition>
  </div>
</template>
<script>
export default {
  name: 'Combobox',
  data () {
    return {
      open: false,
      selected: ''
    }
  },
  props: {
    options: {
      type: Array,
      required: true
    },
    type: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    typeBackground () {
      if (!this.type) {
        return 'default'
      } else {
        return 'forms'
      }
    },
    typeDisabled () {
      if (!this.disabled) {
        return ''
      } else {
        this.selected = this.options[0].title
        return 'disabled'
      }
    }
  },
  created () {
    this.selected = this.options[0].title
  },
  methods: {
    setSelectedValue (option) {
      this.selected = option.title
      this.open = false
      this.$emit('setOptionValue', option.value)
    }
  }
}
</script>
<style lang="scss" scoped>
@import "../assets/variables.scss";

.select-options {
  &__select {
    list-style-type: none;
    margin: 0 0 2rem 0;
    padding: 0;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.24);
    li:first-child {
      display: none;
    }
  }
  button {
    height: 48px;
    width: 100%;
    text-align: left;
    cursor: pointer;
    border: 0;
    font-family: $grot;
    font-size: 1rem;
    padding-left: 1.75rem;
    transition: background-color ease-in-out 0.35s;
    &.default {
      background-color: rgba(255, 255, 255, 0.05);
      color: #fff;
      &:hover {
        background-color: rgba(255, 255, 255, 0.15);
      }
    }
    &.disabled {
      cursor: not-allowed;
      background-color: #ccc;
      &:hover {
        background-color: #ccc;
      }
    }
    &.forms {
      background-color: rgba(0, 0, 0, 0.08);
      color: #000;
      &:hover {
        background-color: rgba(0, 0, 0, 0.1);
      }
    }
    &:focus {
      outline: none;
    }
  }
  &__toggler {
    display: flex;
    align-items: center;
    justify-content: space-between;
    border-radius: 20px;
    margin-bottom: 10px;
    .icon {
      font-size: 40px;
      transition: transform ease-in-out 0.15s;
      &.open {
        transform: rotate(180deg);
      }
    }
  }
}
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.25s;
}
.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>
