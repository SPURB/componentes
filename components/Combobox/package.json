{
  "name": "@spurb/combobox",
  "version": "0.0.0",
  "publishConfig": {
    "access": "public"
  },
  "main": "dist/Combobox.commons.js",
  "scripts": {
    "build": "vue-cli-service build --target lib --formats commonjs --name Combobox ./Combobox.vue"
  }
}
