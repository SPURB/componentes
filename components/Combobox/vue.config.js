const { configureWebpack, css } = require('../vue.config.common')
module.exports = { configureWebpack, css }
