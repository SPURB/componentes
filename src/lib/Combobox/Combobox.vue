<template>
  <div class="select-options">
    <button
      class="select-options__toggler"
      :class="[typeBackground, typeDisabled]"
      @click.prevent="open = !open"
      :disabled="disabled"
    >
      <div>{{ selected }}</div>
      <i class="icon icon-expandir" :class="{ open }"></i>
    </button>
    <transition name="fade">
      <ul class="select-options__select" v-show="open">
        <li
          class="select-options__options"
          v-for="(option, index) in options"
          :value="option.value"
          :key="index"
        >
          <button
            @click.prevent="setSelectedValue(option)"
          >
            {{ option.title }}
          </button>
        </li>
      </ul>
    </transition>
  </div>
</template>
<script>
export default {
  name: 'Combobox',
  data () {
    return {
      open: false,
      selected: ''
    }
  },
  props: {
    options: {
      type: Array,
      required: true
    },
    type: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    typeBackground () {
      if (!this.type) {
        return 'default'
      } else {
        return 'forms'
      }
    },
    typeDisabled () {
      if (!this.disabled) {
        return ''
      } else {
        this.selected = this.options[0].title
        return 'disabled'
      }
    }
  },
  created () {
    this.selected = this.options[0].title
  },
  methods: {
    setSelectedValue (option) {
      this.selected = option.title
      this.open = false
      this.$emit('setOptionValue', option.value)
    }
  }
}
</script>
<style scoped>
.select-options__select {
  list-style-type: none;
  margin: 0 0 2rem 0;
  padding: 0;
  box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.24);
}
.select-options__select li:first-child {
  display: none;
}
.select-options button {
  height: 48px;
  width: 100%;
  text-align: left;
  cursor: pointer;
  border: 0;
  font-family: 'Roboto', 'Segoe UI', 'Helvetica', Arial, sans-serif;
  font-size: 1rem;
  padding-left: 1.75rem;
  transition: background-color ease-in-out 0.35s;
}
.select-options button.default {
  background-color: rgba(255, 255, 255, 0.05);
  color: #fff;
}
.select-options button.default:hover {
  background-color: rgba(255, 255, 255, 0.15);
}
.select-options button.disabled {
  cursor: not-allowed;
  background-color: #ccc;
}
.select-options button.disabled:hover {
  background-color: #ccc;
}
.select-options button.forms {
  background-color: rgba(0, 0, 0, 0.08);
  color: #000;
}
.select-options button.forms:hover {
  background-color: rgba(0, 0, 0, 0.1);
}
.select-options button:focus {
  outline: none;
}
.select-options__toggler {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-radius: 20px;
  margin-bottom: 10px;
}
.select-options__toggler .icon {
  font-size: 40px;
  transition: transform ease-in-out 0.15s;
}
.select-options__toggler .icon.open {
  transform: rotate(180deg);
}
.fade-enter-active, .fade-leave-active {
  transition: opacity 0.25s;
}
.fade-enter, .fade-leave-to {
  opacity: 0;
}
</style>
