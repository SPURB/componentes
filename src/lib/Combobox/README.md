# Combobox.vue
Componente de dropdown para seleção de items

## Exemplo
https://spurb.github.io/componentes/?path=/story/combobox--normal

## Instruções de uso
Declare
```js
import { Combobox } from '@spurb/componentes'
export default {
  components: { Combobox }
}
```

Exemplo de uso no template
```html
<combobox
  :disabled="true"
  :open="true"
  fill-brand="#038375"
  :options="[
      {
        title: 'Selecione um projeto',
        value: 'valor default'
      },
      {
        title: 'PIU Bairros do Tamandatueí',
        value: 'projeto-1'
      },
      {
        title: 'PIU Jurubatuba',
        value: 'projeto-2'
      },
      {
        title: 'PIU Setor Central',
        value: 'projeto-3'
      }
  ]"
  @setOptionValue="metodoComponentePai"
/>
```

### Props
| nome | data type | descrição |
|:-|:-|:-|
| `fill-type` |  String | Cor de preenchimento do texto |
| `fill-brand` | String | Cor de preenchimento da marca |
| `opacity` | Number | Define a opacidade do elemento número variando de 0 a 1 |


### Events
| nome | Parâmetro do callback |
|:-|:-|
| `setOptionValue` | Valor de `value` da opção selecionada pelo usuário|
