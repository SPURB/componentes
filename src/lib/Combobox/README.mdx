instalar
`npm i @spurb/combobox`

exemplo:
https://spurb.github.io/componentes/?path=/story/combobox--normal
