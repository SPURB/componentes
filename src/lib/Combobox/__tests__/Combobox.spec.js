import { shallowMount } from '@vue/test-utils'
import Combobox from '../Combobox.vue'

const options = [
  {
    title: 'Selecione um projeto',
    value: 0
  },
  {
    title: 'PIU Bairros do Tamandatueí',
    value: 1
  }
]

describe('Combobox.vue', () => {
  it('Renderiza corretamente passando props', () => {
    const wrapper = shallowMount(Combobox, {
      propsData: { options }
    })
    expect(wrapper.element).toMatchSnapshot() // compara renderização por snashot
  })
})
