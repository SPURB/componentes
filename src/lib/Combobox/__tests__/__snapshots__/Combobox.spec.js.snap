// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox.vue Renderiza corretamente passando props 1`] = `
<div
  class="select-options"
>
  <button
    class="select-options__toggler default"
  >
    <div>
      Selecione um projeto
    </div>
     
    <i
      class="icon icon-expandir"
    />
  </button>
   
  <transition-stub
    name="fade"
  >
    <ul
      class="select-options__select"
      style="display: none;"
    >
      <li
        class="select-options__options"
        value="0"
      >
        <button>
          
          Selecione um projeto
        
        </button>
      </li>
      <li
        class="select-options__options"
        value="1"
      >
        <button>
          
          PIU Bairros do Tamandatue√≠
        
        </button>
      </li>
    </ul>
  </transition-stub>
</div>
`;
